### ISSUE TITLE
Headless OpenGL rendering

### ISSUE THREAD

**Issue by qianyizh** [IntelVCL/Open3D#17]
This can be useful in scenarios where users want to render a (long) video.
It can also be useful when Open3D is deployed in a remote server and window creation is a problem.

Some information regarding GLFW: https://github.com/glfw/glfw/issues/648

**Comment by bhack** [comment]
Other than VNC there is also xpra https://xpra.org/trac/wiki/Usage/OpenGL

**Comment by rukie** [comment]
As another user, this would be an excellent feature! 👍

**Comment by syncle** [comment]
Working on this. I am using glfw3.3_dev to avoid hassles.

**Comment by takanokage** [comment]
Another approach is to use Docker to create a headless environment for Open3D.
I already have a docker image based on ubuntu 16.04 with xvfb and x11vnc that works. I've also tested one light window manager (ratpoison) but I'm not too happy with it and want to test others. Next I need to test building/running Open3D inside it.

If we need access to the host GPU (NVIDIA) there are already good solutions to get that: NVIDIA-docker and NVIDIA docker images with CUDA and cuDNN.

I've also been looking into using Alpine as the base if the image size is a concern. The Ubuntu image with xvfb and x1vnc gets to about 450MB while the Alpine image to just about 70MB.

**Comment by qianyizh** [comment]
The purpose is to make Open3D rendering part running on cloud (AWS/GCloud). @syncle can you help test the solution on GCloud?

**Comment by syncle** [comment]
Sure. I can use a fresh Ubuntu machine running on GCloud. @takanokage, can you elaborate your steps for me? That will be used for documents too.

**Comment by takanokage** [comment]
I will post more once I have it working.
I'm at the point where I can build Open3D inside the docker container. However the verification step after the build fails, for this reason I added [issue 241](https://github.com/IntelVCL/Open3D/issues/241).

**Comment by takanokage** [comment]
I'm getting the same `X11: RandR gamma ramp support seems broken` error.
Next: I'll check the `VirtualGL` suggestion by [Qianyi](https://github.com/IntelVCL/Open3D/issues/17#issuecomment-339391580).

$ cd ~/Open3D/build/lib/Tutorial/Basic
$ python3 rgbd_redwood.py 
press q

```
Read Redwood dataset
RGBDImage of size 
Color image : 640x480, with 1 channels.
Depth image : 640x480, with 1 channels.
Use numpy.asarray to access buffer data.
GLFW Error: X11: RandR gamma ramp support seems broken
GLFW Error: Linux: Failed to watch for joystick connections in /dev/input: No such file or directory
GLFW Error: Linux: Failed to open joystick device directory /dev/input: No such file or directory
```

**Comment by syncle** [comment]
From @qianyizh's post

> There is a risk that you will have this error
```
glfw.GLFWError: (65544) b'X11: RandR gamma ramp support seems broken'
```
> This is an error of version 3.2 and older. They don't support headless rendering. Since 3.3 (current master) it is supported. Then everything should be fine.

It might be good to try Open3D + GLFW3.3_dev solution via
```
git clone -b headless_rendering https://github.com/syncle/Open3D.git
```

**Comment by takanokage** [comment]
Jaesik, I just tired your branch with [glfw 3.3](https://github.com/syncle/Open3D/tree/headless_rendering) inside my docker container.
Somehow I'm still getting the same `X11: RandR gamma ramp support seems broken` error.
Do I need to do anything special besides getting your branch?

**Comment by syncle** [comment]
Yes. Try it with
```
cmake -DOpen3D_USE_NATIVE_DEPENDENCY_BUILD=OFF ../src
```

**Comment by takanokage** [comment]
Thanks!
So that did something:
GLFW Error: OSMesa: Library not found
Failed to create window
[DrawGeometries] Failed creating OpenGL window.

Did you install a special package for this?

**Comment by syncle** [comment]
I think this should work
```
sudo apt-get install libgl1-mesa-dev -y 
sudo apt-get install libosmesa6-dev -y
```

**Comment by takanokage** [comment]
This is the output of `glxinfo`: `OpenGL version string: 3.0 Mesa 17.2.8`

`libgl1-mesa-dev` was already installed.
I installed `libosmesa6-dev` and the previous glfw error dissapeared.

Now I'm getting a segmentation fault when running Tutorial/Basic/rgbd-redwood.py. I'll troubleshoot and let you know what I found.

**Comment by takanokage** [comment]
After going trough lots of resources it would seem the segmentation fault is likely an issue introduced by the use of docker.
Among the things I have tried:
- use an NVIDIA driver+CUDA based image
- use nvidia-docker to run the containers

Others have seen the same error however there's no consensus on how to address it.

**Comment by takanokage** [comment]
The `Segmentation fault` went away. I successfully ran `Tutorial/Basic/rgbd_redwood.py` and 
`Tutorial/Advanced/customize_visualization.py`.

I cloned and tested using the [master](https://github.com/IntelVCL/Open3D/tree/master) branch rather than the [headless_rendering](https://github.com/syncle/Open3D/tree/headless_rendering) branch with `glfw3.3`.

I'm still getting the `X11: RandR gamma ramp support seems broken` error but that shows that this error has a different cause/meaning than we previously thought.

**Comment by qianyizh** [comment]
You are not linking to the dev branch of glfw (3.3).
In dev branch this message has been removed ``X11: RandR gamma ramp support seems broken``. If you are using the correct version, you should not be able to see it.
Read the code of glfw, you will see.

**Comment by takanokage** [comment]
In other words this is more like a warning instead of an error and will go away once we move to glfw 3.3?

**Comment by qianyizh** [comment]
https://github.com/glfw/glfw/blame/6e69f63ffa4059d0f16344e6b038285c1ca78c16/src/x11_init.c
https://github.com/glfw/glfw/commit/b0258803c6d6e8b03f194a73457a1e0645196f1a#diff-9c6aefaea564bfbb5f29c9b9026e1b34
https://github.com/glfw/glfw/commit/5fe4dfb511e12b954f9fba11d989df48f5f203fd

**Comment by takanokage** [comment]
1. If I clone/build/run the [headless_rendering](https://github.com/syncle/Open3D/tree/headless_rendering) branch I get the segmentation fault.
`$ cd ~/Open3D/build/lib/Tutorial/Advanced`
`$ python3 customized_visualization.py`

> Reading PLY: [========================================] 100%
> 1. Customized visualization to mimic DrawGeometry
> Segmentation fault (core dumped)

2. If I clone/build/run the [master](https://github.com/IntelVCL/Open3D/tree/master) branch I get the `RandR` and `joystick` errors but the tutorials run fine.
`$ cd ~/Open3D/build/lib/Tutorial/Advanced`
`$ python3 customized_visualization.py`

> Reading PLY: [========================================] 100%
> 1. Customized visualization to mimic DrawGeometry
> GLFW Error: X11: RandR gamma ramp support seems broken
> GLFW Error: Linux: Failed to watch for joystick connections in /dev/input: No such file or directory
> GLFW Error: Linux: Failed to open joystick device directory /dev/input: No such file or directory
> 2. Customized visualization with a rotating view
> GLFW Error: X11: RandR gamma ramp support seems broken
> GLFW Error: Linux: Failed to watch for joystick connections in /dev/input: No such file or directory
> GLFW Error: Linux: Failed to open joystick device directory /dev/input: No such file or directory
> 3. Customized visualization showing normal rendering
> GLFW Error: X11: RandR gamma ramp support seems broken
> GLFW Error: Linux: Failed to watch for joystick connections in /dev/input: No such file or directory
> GLFW Error: Linux: Failed to open joystick device directory /dev/input: No such file or directory
> 4. Customized visualization with key press callbacks
>    Press 'K' to change background color to black
>    Press 'R' to load a customized render option, showing normals
>    Press ',' to capture the depth buffer and show it
>    Press '.' to capture the screen and show it
> GLFW Error: X11: RandR gamma ramp support seems broken
> GLFW Error: Linux: Failed to watch for joystick connections in /dev/input: No such file or directory
> GLFW Error: Linux: Failed to open joystick device directory /dev/input: No such file or directory
> 5. Customized visualization playing a camera trajectory
> GLFW Error: X11: RandR gamma ramp support seems broken
> GLFW Error: Linux: Failed to watch for joystick connections in /dev/input: No such file or directory
> GLFW Error: Linux: Failed to open joystick device directory /dev/input: No such file or directory
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00000
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00001
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00002
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00003
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00004
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00005
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00006
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.
> Capture image 00007
> [ViewControl] ConvertFromPinholeCameraParameters() failed because window height and width do not match.

**Comment by syncle** [comment]
> After going trough lots of resources it would seem the segmentation fault is likely an issue introduced by the use of docker.

If the issue is due to docker, how about not using docker? You can make virtual machine instead. As @qianyizh have tested on Ubuntu systems, we just want to finalize this issue with GLFW3.3, instead of blocked by new issues regarding the docker.

**Comment by takanokage** [comment]
Actually that was a speculation on my part.
A later experiment showed a correlation between glfw 3.3 and the presence of the segmentation fault. The master branch works fine with docker.

**Comment by takanokage** [comment]
I have headless rendering working inside docker with the master branch of Open3D.
Next I'll be working on the documentation before submitting a pull request.

**Comment by takanokage** [comment]
**Bind mounts** (#254)
I propose we make the whole Open3D folder available both inside the docker container and on the host side. In this way we not only gain direct access to the images generated but we also enable developing Open3D on the host side and testing on both host side and docker side of things.

**Comment by takanokage** [comment]
Because we moved #254 under #17 I need to postpone the pull request until after I implement/test it.

**Comment by takanokage** [comment]
I got headless working together with bind mounting.
Bind mounting was a ton of work and should've been an independent pull request.

**Comment by takanokage** [comment]
Usage notes:
1. clone my [branch](https://github.com/takanokage/Open3D/tree/issue_%230017)
2. $ `cd <Open3D path>/utilities/docker/ubuntu-xvfb/tools`
3. $ `./build.sh` to build the image
4. $ `./attach.sh` to clone/build the master Open3D repo and start the Open3D container
5. $ `./headless_sample.sh` to run the sample that renders some images and saves them to disk.
The sample will render some images which can be accessed on the host at ~/Open3D_docker. These files don't go away if the container is stopped.

Notes:
- the sample will not return. Ctrl+c to exit. Need to update the sample to exit on it's own.
- the cloning is done for now to ~/Open3D_docker. We have options: find a way to let the user specify the destination or just reuse the current location of Open3D.
- TODO: uncomment entries in the dependencies section inside the Dockerfile.

**Comment by syncle** [comment]
Addressed in #273.
